### PR TITLE
feat(api): support both array and string for the "content" field

### PR DESCRIPTION
**Pull request by kkaneda**
- Add the new "content" field (= different proto field ID). The type of the field is an array.
- If the "content" contains a string value in the request JSON, convert it to an array before proto unmarshalling.
- Rename the current "content" field to the "legacy_content", but continue to set the value. This is requierd for backward compatibility.

### CODE REVIEW

File: api/v1/inference_server.proto [hunk 1]
```diff
@@ -28,25 +21,29 @@ message CreateChatCompletionRequest {
...
-      ContentType type = 1;
+      // Refer to https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages
+      message ImageUrl {
+	string url = 1;
+	string detail = 2;
+      }
+      message InputAudio {
+	string data = 1;
+	string format = 2;
+      }
```
Comment (Ladicle) [inline api/v1/inference_server.proto:33]: nits: it seems like the indent is off.

File: server/internal/server/legacy_completions.go [hunk 2]
```diff
@@ -222,8 +222,8 @@ func toCreateChatCompletionRequest(req *v1.CreateCompletionRequest) *v1.CreateCh
 	return &v1.CreateChatCompletionRequest{
 		Messages: []*v1.CreateChatCompletionRequest_Message{
 			{
-				Content: req.Prompt,
-				Role:    "user",
+				//Content: []string{req.Prompt},
```
Comment (Ladicle) [inline server/internal/server/legacy_completions.go:225]: maybe
```suggestion
```

File: server/internal/server/marshal_test.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	assert.NoError(t, err)
+
+	/*
+
+
+		var req v1.CreateChatCompletionRequest
+		err := json.Unmarshal([]byte(body), &req)
+		assert.NoError(t, err)
+		assert.Equal(t, "You are a helpful assistant.", req.Messages[0].Content)
+	*/
```
Comment (Ladicle) [inline server/internal/server/marshal_test.go:45]:
```suggestion
```

File: triton-proxy/internal/server/triton.go [hunk 4]
```diff
@@ -18,7 +18,10 @@ func buildEnsembleGenerateRequest(req *v1.CreateChatCompletionRequest) *ensemble
 	// this to a template.
 	input := "<|begin_of_text|>"
 	for _, m := range req.Messages {
-		input += fmt.Sprintf("<|start_header_id|> %s <|end_header_id|>\n %s '\n<|eot_id|>\n", m.Role, m.Content)
+		for _, c := range m.Content {
+			// TODO(kenji): Support other types.
+			input += fmt.Sprintf("<|start_header_id|> %s <|end_header_id|>\n %s '\n<|eot_id|>\n", m.Role, c.Text)
```
Comment (Ladicle) [inline triton-proxy/internal/server/triton.go:23]: This has not been changed in this PR, but is the `'` before `eot_id` a template syntax?

**Review by Ladicle — APPROVED** [review]
LGTM

File: server/internal/server/marshal_test.go (hunk 3 above)
Comment (kkaneda) [reply server/internal/server/marshal_test.go:45]: my bad. let me delete this file.

File: triton-proxy/internal/server/triton.go (hunk 4 above)
Comment (kkaneda) [reply triton-proxy/internal/server/triton.go:23]: nice catch! this looks like a bug. let me create a separate PR to fix.

File: api/v1/inference_server.proto (hunk 1 above)
Comment (kkaneda) [reply api/v1/inference_server.proto:33]: ah confusing, but I've moved these messages inside `Content` to follow the pattern in this file.

**Review by guangrui-cloudnatix — APPROVED** [review]

File: server/internal/server/completions.go [hunk 5]
```diff
@@ -67,12 +67,26 @@ func (s *S) CreateChatCompletion(
...
+		return
+	}
+
 	// TODO(kenji): Use runtime.JSONPb from github.com/grpc-ecosystem/grpc-gateway/v2.
 	// That one correctly handles the JSON field names of the snake case.
 	if err := json.Unmarshal(reqBody, &createReq); err != nil {
 		httpError(w, err.Error(), http.StatusInternalServerError, &usage)
 		return
 	}
+	// Set the LegacyContext field for the backward compatibility.
```
[the remark this answers is not in the document]
Comment (kkaneda) [reply server/internal/server/completions.go:82]:
```suggestion
	// Set the LegacyContent field for the backward compatibility.
```

**Comment by kkaneda** [comment]
Tested locally with Ollama.  Confirmed that the server with this change works with an older version of engine.

**Pull request merged**